### ISSUE TITLE
aws_lambda_function - change of layer arn should trigger new version

### ISSUE THREAD

**Issue by ben-nz** [terraform-providers/terraform-provider-aws#9707]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v0.12.6
+ provider.aws v2.23.0

### Affected Resource(s)

* aws_lambda_function

### Terraform Configuration Files

```hcl
resource "aws_lambda_function" "lambda" {
  function_name    = "${var.full_env_name}-${var.name}"
  filename         = local.lambda_zip_path
  source_code_hash = filebase64sha256(local.lambda_zip_path)
  handler          = "index.${var.name}"
  runtime          = "nodejs8.10"
  publish          = true
  role             = var.lambda_exec_role_arn
  timeout          = 900
  layers           = var.layers

  tracing_config {
    mode = "Active"
  }

  environment {
    variables = var.env_vars
  }

  tags = var.lambda_tags
}
```

### Terraform Plan Output
```
 module.lambda_subscribe_to_mac_updates.aws_lambda_function.lambda will be updated in-place
  ~ resource "aws_lambda_function" "lambda" {
        arn                            = "arn"
        filename                       = "../publish/<lambda>.zip"
        function_name                  = "<name>"
        handler                        = "index.<handler>"
        id                             = "<id>"
        invoke_arn                     = "arn:aws:apigateway:us-east-1:lambda:path/2015-03-31/functions/arn:aws:lambda:<function>/invocations"
        last_modified                  = "2019-08-09T15:13:02.298+0000"
      ~ layers                         = [
          - "<old layer arn>",
        ] -> (known after apply)
        memory_size                    = 128
        publish                        = true
        qualified_arn                  = "<arn>"
        reserved_concurrent_executions = -1
        role                           = "<role arn>"
        runtime                        = "nodejs8.10"
        source_code_hash               = "<hash>"
        source_code_size               = 6342
        timeout                        = 900
        version                        = "9"

    }
```

### Expected Behavior
When the layer arn is changed for a lambda and publish = true, a new version of the lambda should be created

### Actual Behavior
No new version of the lambda is created, the modified lambda is only available at $LATEST.

### Steps to Reproduce

1. Create lambda with a layer attached and publish = true
2. Change the layer, triggering a new layer version and therefore new layer arn
3. Observe on `terraform plan` that the lambda version stays the same even though the layer arn changes

### References

Logic for deciding to publish a new lambda version:
https://github.com/terraform-providers/terraform-provider-aws/blob/cc3dba0b108f7a28603da2bb0bf32695f5aabe5e/aws/resource_aws_lambda_function.go#L261-L271

### Remarks
Hi all, we have recently switched to using lambda layers for an existing terraform stack of nodejs 8.10 functions. We are using a layer for shared npm packages. We use an alias attached to the latest published version of our lambdas as a target for our API gateways. Not relying on $LATEST for the lambdas allows us to do rollbacks quickly and easily.

I have noticed that modifying a layer attached to our lambdas does not trigger a new version as it it does not constitute a code change according to the provider logic (needsFunctionCodeUpdate returns false). I would argue that changes to layer code is indeed a 'code change', it's just that the code change is behind a layer of abstraction. According to the [AWS docs](https://docs.aws.amazon.com/lambda/latest/dg/versioning-intro.html):

> When you publish a version, AWS Lambda makes a snapshot copy of the Lambda function code (and configuration) in the $LATEST version. A published version is immutable. That is, you can't change the code or configuration information.

In the interests of immutability, I suggest that a change of layer arns should be an additional trigger for publishing a new lambda version. I'm new to golang but am happy to take a shot at writing a PR if people are in agreement. Thanks!

**Comment by bflad** [comment]
The fix for this has been merged and will release with version 2.50.0 of the Terraform AWS Provider, later today. Thanks to @nemreid for the implementation. 👍 

Please note that this resource will now use a separate `PublishVersion` API call, so any restrictive IAM policies for the Terraform run may require updates.

**Comment by ben-nz** [comment]
Many thanks to @nemreid and @bflad :-)
